### PR TITLE
Fix Issue 18728 - make fdim handle nan

### PR DESCRIPTION
**Pull request by Andu033**
Please see this issue discussion :https://issues.dlang.org/show_bug.cgi?id=18728

### CODE REVIEW

File: std/math.d [hunk 1]
```diff
@@ -7337,16 +7337,21 @@ T nextafter(T)(const T x, const T y) @safe pure nothrow @nogc
  *      $(TR $(TD x $(LT)= y) $(TD +0.0))
  *      )
  */
-real fdim(real x, real y) @safe pure nothrow @nogc { return (x > y) ? x - y : +0.0; }
+real fdim(real x, real y) @safe pure nothrow @nogc
+{   
+    if(isNaN(x) || isNaN(y))
+        return real.nan;
+    return (x > y) ? x - y : +0.0;
```
Comment (n8sh) [inline std/math.d]: How about
```d
if (x > y) return x - y;
else if (x <= y) return +0.0;
else return real.nan;
```

The output with `ldc2 -O2` seems better:
https://run.dlang.io/is/Q5YlQZ
<details>
<summary>fdim_impl2 (suggested version)</summary>

```
fdim_impl2:
	.cfi_startproc
	fldt	24(%rsp)
	fldt	8(%rsp)
	fucomi	%st(1)
	jbe	.LBB2_2
	fsubp	%st(1)
	retq
.LBB2_2:
	fxch	%st(1)
	fucompi	%st(1)
	fstp	%st(0)
	fldz
	flds	.LCPI2_0(%rip)
	fcmovnb	%st(1), %st(0)
	fstp	%st(1)
	retq
.Lfunc_end2:
	.size	fdim_impl2, .Lfunc_end2-fdim_impl2
	.cfi_endproc
```
</details>

versus

<details>
<summary>fdim_impl1 (isNaN(x) || isNaN(y) version)</summary>

```
fdim_impl1:
	.cfi_startproc
	fldt	8(%rsp)
	fld	%st(0)
	fstpt	-24(%rsp)
	movl	-16(%rsp), %ecx
	notl	%ecx
	fldt	24(%rsp)
	movabsq	$9223372036854775807, %rax
	flds	.LCPI0_0(%rip)
	testl	$32767, %ecx
	jne	.LBB0_4
	movq	-24(%rsp), %rcx
	andq	%rax, %rcx
	je	.LBB0_4
	fstp	%st(2)
	fstp	%st(0)
	jmp	.LBB0_3
.LBB0_4:
	fld	%st(1)
	fstpt	-40(%rsp)
	movl	-32(%rsp), %ecx
	notl	%ecx
	testl	$32767, %ecx
	fxch	%st(2)
	fxch	%st(1)
	jne	.LBB0_7
	andq	-40(%rsp), %rax
	jne	.LBB0_6
.LBB0_7:
	fstp	%st(2)
	fucomi	%st(1)
	fsubp	%st(1)
	fldz
	fcmovnbe	%st(1), %st(0)
	fstp	%st(1)
.LBB0_3:
	fldz
	fldz
.LBB0_6:
	fstp	%st(1)
	fstp	%st(0)
	retq
.Lfunc_end0:
	.size	fdim_impl1, .Lfunc_end0-fdim_impl1
	.cfi_endproc
```
</details>

Comment (n8sh) [reply std/math.d]: (We can't do `return !(x < y) ? x - y : +0.0` because although it correctly handles NaN it doesn't correctly handle both x & y being same-signed infinity.)

File: std/math.d [hunk 2]
```diff
@@ -7337,16 +7337,21 @@ T nextafter(T)(const T x, const T y) @safe pure nothrow @nogc
  *      $(TR $(TD x $(LT)= y) $(TD +0.0))
```
Comment (n8sh) [inline std/math.d]: The special values table will need to be updated.

**Comment by JackStouffer** [comment]
Ping @Andu033

File: std/math.d [hunk 3]
```diff
@@ -7697,9 +7702,7 @@ if (isFloatingPoint!(F) && isFloatingPoint!(G))
                 }
             }
             else
-            {
-                if (y > 0.0)
-                    return F.infinity;
+            { if (y > 0.0) return F.infinity;
```
Comment (ibuclaw) [inline std/math.d]: Unrelated change, and other review comments still need to be addressed.

**Comment by RazvanN7** [comment]
This issue seems to have been fixed by: https://github.com/dlang/phobos/pull/6896
